Support custom forms in native user tasks

tmetzke opened the pull request:
## Description

<!-- Please explain the changes you made here. -->

## Related issues

<!-- Which issues are closed by this PR or are related -->

closes #16006 

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

ana-vinogradova-camunda reviewed (approved):
👍🏻 LGTM, I really like the detailed tests, thank you

koevskinikola commented on protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/usertask/UserTaskRecord.java:
```diff
@@ -70,14 +72,15 @@ public final class UserTaskRecord extends UnifiedRecordValue implements UserTask
 
   public UserTaskRecord() {
-    super(16);
+    super(17);
```
:wrench: We'll need to adjust this number depending if https://github.com/camunda/zeebe/pull/15915#pullrequestreview-1862549968 is already merged.

tmetzke replied:
Yep, I hate PRs that depend on each other 😆

tmetzke replied:
Done after rebasing onto `main`

koevskinikola commented on bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/builder/AbstractUserTaskBuilder.java:
```diff
@@ -156,16 +156,20 @@ public B zeebeFormId(final String formId) {
   }
 
+  @Override
+  public B zeebeExternalReference(final String externalReference) {
```
:upside_down_face: ~Same as in the `UserTaskRecord`, would this method be more readable if it was named `zeebeExternalFormReference`?~

Nvm, I just saw that it's an attribute of the `formDefinition` element, so it provides context and the above isn't necessary.

tmetzke replied:
Yeah, but the `formId` and `formKey` are also not named `id` or `key`, so 🤷

tmetzke replied:
Renamed it in the builder at least (not in the model itself).

koevskinikola commented on bpmn-model/src/test/java/io/camunda/zeebe/model/bpmn/validation/ZeebeTaskValidatorFormTest.java:
:+1: Nice additional coverage :smile:

koevskinikola reviewed (approved):
LGTM @tmetzke :tada: 

I left some small nitpicks, but they don't impact the correctness od the code.

tmetzke commented on protocol/src/main/java/io/camunda/zeebe/protocol/record/value/UserTaskRecordValue.java:
```diff
@@ -44,6 +44,8 @@ public interface UserTaskRecordValue
 
   long getFormKey();
 
+  String getExternalReference();
```
Yeah, that's a really good point. That might make more sense. Then again, it doesn't have to be a form at all but could be an application as well. I don't mind either way. @christian-konrad, any preferences on this naming? Should we rather call it `externalFormReference` than `externalReference`? We would need to consistently do this then here, in t modeling, and the REST API later on I think.

tmetzke replied:
I have addressed this issue in the API doc as well to sync with @christian-konrad and @aleksander-dytko. I will add the decisions here for transparency.

tmetzke replied:
I adjusted this in the engine itself at least because it's needed to provide enough context. The modeling aspect still needs to be clarified.

tmetzke replied:
I'll go ahead and merge this as is for now. Should the modeling attribute change, we can create a follow-up issue.

tmetzke commented:
Updates:
* Rebased onto `main`
* Renamed the external reference attribute in the engine (and related components but not in the model itself yet)
* Fixed a bug with expression handling for the external reference
